src/lib.rs

```diff
@@ -210,54 +212,63 @@ impl Core {
                 );
                 slots.get_unchecked(idx)
             };
-            let state = test_dbg!(slot.state.load(Acquire));
-
+            let state = test_dbg!(slot.state.load(SeqCst));
+            // slot is writable
             if test_dbg!(state == tail) {
-                // Move the tail index forward by 1.
                 let next_tail = self.next(idx, gen);
+                // try to advance the tail
                 match test_dbg!(self
                     .tail
                     .compare_exchange_weak(tail, next_tail, SeqCst, Acquire))
                 {
                     Ok(_) => {
-                        // We got the slot! It's now okay to write to it
-                        test_println!("claimed tail slot [{}]", idx);
-                        // Claim exclusive ownership over the slot
-                        let ptr = slot.value.get_mut();
-
-                        // Initialize or recycle the element.
-                        unsafe {
-                            // Safety: we have just claimed exclusive ownership over
-                            // this slot.
-                            let ptr = ptr.deref();
-                            if gen == 0 {
-                                ptr.write(recycle.new_element());
-                                test_println!("-> initialized");
-                            } else {
-                                // Safety: if the generation is > 0, then the
-                                // slot has already been initialized.
-                                recycle.recycle(ptr.assume_init_mut());
-                                test_println!("-> recycled");
+                        test_println!("advanced tail {} to {}", tail, next_tail);
+                        test_println!("claimed slot [{}]", idx);
+                        let has_reader = test_dbg!(slot.has_reader.load(SeqCst));
```

Comment:
it looks like it was necessary to change these loads from `Acquire` to `SeqCst` because we need the loads of the `state` and `has_reader` fields to have a happens-before relationship? if we apply my above suggestion about merging the has-reader bit into the `state` variable, we could avoid the need to synchronize between two loads, and we could still perform `Acquire` loads here. this might improve performance a bit...